test/upgradeableProxy/Dispatcher.upgrade.t.sol

```diff
@@ -0,0 +1,353 @@
+    IbcEndpoint dest;
+    bytes payload = bytes("msgPayload");
+    bytes appAck = abi.encodePacked('{ "account": "account", "reply": "got the message" }');
+
+    function setUp() public override {
+        super.setUp();
+        string memory marsPort = string(abi.encodePacked(portPrefix, getHexBytes(address(mars))));
+        dest = IbcEndpoint(marsPort, channelId);
+    }
+
+    function test_success() public {
+        for (uint64 index = 0; index < 3; index++) {
+            uint64 packetSeq = index + 1;
+            vm.expectEmit(true, true, true, true, address(dispatcher));
+            emit RecvPacket(address(mars), channelId, packetSeq);
+            vm.expectEmit(true, true, false, true, address(dispatcher));
+            emit WriteAckPacket(address(mars), channelId, packetSeq, AckPacket(true, appAck));
+            dispatcher.recvPacket(
+                IbcReceiver(mars), IbcPacket(src, dest, packetSeq, payload, ZERO_HEIGHT, maxTimeout), validProof
+            );
+        }
+    }
+
+    // recvPacket emits a WriteTimeoutPacket if timestamp passes chain B's block time
+    function test_timeout_timestamp() public {
+        uint64 packetSeq = 1;
+        IbcPacket memory pkt = IbcPacket(src, dest, packetSeq, payload, ZERO_HEIGHT, 1);
+        vm.expectEmit(true, true, true, true, address(dispatcher));
+        emit RecvPacket(address(mars), channelId, packetSeq);
+        vm.expectEmit(true, true, false, true, address(dispatcher));
+        emit WriteTimeoutPacket(address(mars), channelId, packetSeq, pkt.timeoutHeight, pkt.timeoutTimestamp);
+        dispatcher.recvPacket(IbcReceiver(mars), pkt, validProof);
+    }
+
+    // recvPacket emits a WriteTimeoutPacket if block height passes chain B's block height
+    function test_timeout_blockHeight() public {
+        uint64 packetSeq = 1;
+        IbcPacket memory pkt = IbcPacket(src, dest, packetSeq, payload, Height(0, 1), 0);
+        vm.expectEmit(true, true, true, true, address(dispatcher));
+        emit RecvPacket(address(mars), channelId, packetSeq);
+        vm.expectEmit(true, true, false, true, address(dispatcher));
+        emit WriteTimeoutPacket(address(mars), channelId, packetSeq, pkt.timeoutHeight, pkt.timeoutTimestamp);
+        dispatcher.recvPacket(IbcReceiver(mars), pkt, validProof);
+    }
+
+    // cannot receive packets out of order for ordered channel
+    function test_outOfOrder() public {
+        dispatcher.recvPacket(IbcReceiver(mars), IbcPacket(src, dest, 1, payload, ZERO_HEIGHT, maxTimeout), validProof);
+        vm.expectRevert(abi.encodeWithSignature("unexpectedPacketSequence()"));
+        dispatcher.recvPacket(IbcReceiver(mars), IbcPacket(src, dest, 3, payload, ZERO_HEIGHT, maxTimeout), validProof);
+    }
+
+    // TODO: add tests for unordered channel, wrong port, and invalid proof
+}
+
+// Test Chain A receives an acknowledgement packet from Chain B
+contract DispatcherAckPacketTestSuite is PacketSenderTestBase {
```

Comment:
I'm probably overlooking something in here but where are we testing the dispatcher upgradability?

Reply:
The dispatcher is updated to a new implementation in the setup functions. [Here](https://github.com/open-ibc/vibc-core-smart-contracts/blob/8899883c42798a9b1ddfe36dffeee110712b1be1/test/upgradeableProxy/Dispatcher.upgrade.t.sol#L40-L41) is an example - these two lines are not present in the setup for the non-upgrade tests (i.e. the existing tests before this PR. Since these two lines upgrade the proxy and all tests run after this setup, the testing is done on the upgraded proxy.

Reply:
I got rid of the regression tests and  now added a small test to test for state collisions